### PR TITLE
Review

### PR DESCRIPTION
**Pull request by simonegiacomello**
Abbiamo aperto una nuova branch per la pull request.
Nel file targets.py trovi le funzioni target su cui stiamo lavorando (zdt1, zdt2, zdt3 funzionano, le altre due ancora no probabilmente per problemi sui bounds diversi). Poi abbiamo modificato gli script example_smsego.py e example_nsga2.py in modo tale da poterli eseguire cambiando solo il nome della target function

### CODE REVIEW

**Review by brunoguindani — COMMENTED** [review]
Mi sembra siate sulla buona strada. Non mi piace però la catena di if, oltretutto ripetuta su due main, per descrivere i domini delle funzioni. Potreste mettere in `targets.py` una funzione che riceve l'oggetto funzione, oppure il suo nome, e restituisce quello che vi serve (`f1, f2, PB`). A questo proposito, potrebbe esservi utile il metodo globale `getattr()`. Potete farci cose tipo `getattr(targets, target_name)` per prendere la funzione giusta senza bisogno di catene di if. Ricordatevi di gestire il caso in cui il target richiesto non esiste!

**Comment by simonegiacomello** [comment]
Abbiamo aggiornato la chiamata alle target function spostando tutto nello script targets.py. Non siamo riusciti a fare a meno della catena di if siccome il parametro col nome della funzione viene inserito da riga di comando.
Inoltre stiamo provando a ragionare su come plottare la posterior del GP, però ci sono molti problemi. Il classico grafico dove sulle x ci sono i punti e sulle y le varie funzioni non è applicabile al nostro caso siccome abbiamo più dimensioni sia nei parametri che sulle funzioni obiettivo, quindi abbiamo pensato di plottare il grafico (f1,f2), che però non è per nulla informativo.
Ci sapresti suggerire qualcosa?

**Comment by brunoguindani** [comment]
> Abbiamo aggiornato la chiamata alle target function spostando tutto nello script targets.py. Non siamo riusciti a fare a meno della catena di if siccome il parametro col nome della funzione viene inserito da riga di comando.

Questo non è un problema, fintanto che l'if da aggiungere per scrivere un nuovo target è uno solo in un singolo file. Quindi la vostra soluzione attuale va bene.

> Inoltre stiamo provando a ragionare su come plottare la posterior del GP, però ci sono molti problemi. Il classico grafico dove sulle x ci sono i punti e sulle y le varie funzioni non è applicabile al nostro caso siccome abbiamo più dimensioni sia nei parametri che sulle funzioni obiettivo, quindi abbiamo pensato di plottare il grafico (f1,f2), che però non è per nulla informativo. Ci sapresti suggerire qualcosa?

Visto che c'è un GP *indipendente* per ogni componente della funzione target, farei vari plot ciascuno dei quali ha il singolo GP insieme alla corrispondente componente. Alla fine questo è ciò che succede nel codice. E del resto non mi sembra ci sia molta alternativa per rappresentare una funzione anche solo $\mathbb{R}^2 \to \mathbb{R}^2$

**Comment by simonegiacomello** [comment]
Intuitivamente sarebbe la cosa corretta tenere separati i plot di ogni GP indipendente, ma il problema è che il nostro spazio delle X ha dimensione alta (30 nel caso delle ZDT). Quindi un plot includendo le X non è fattibile.
L'unica cosa che ci pareva sensata era plottare le posterior dei 2 GP in un grafico dove una sta sull'asse x e l'altra sull'asse y, che è quello che intendevo con grafico (f1,f2), ma appunto non sembra essere interpretabile

**Comment by simonegiacomello** [comment]
Buongiorno Bruno, abbiamo abbozzato la funzione per fare il calcolo della posterior del GP. 
Alla fine abbiamo pensato che avendo uno spazio a 30 dimensioni nelle X e 2 nelle Y, il grafico va necessariamente fatto scegliendo una componente di X (abbiamo scelto la 0esima per comodità) e una componente di Y. Nella maggior parte dei nostri target la prima componente di Y è l'identità quindi abbiamo usato la seconda, ma lo script è adattabile a qualunque componente. 

Questa è la reference da cui abbiamo preso spunto: https://scikit-learn.org/stable/auto_examples/gaussian_process/plot_gpr_prior_posterior.html

Il nostro procedimento prevede il sampling di n_samples funzioni dal GP, ciascuna valutata in un insieme di punti costituito dalle osservazioni che ha fatto l'algoritmo (tante quante il numero di iterazioni con cui lo eseguiamo) e da punti generati casualmente (n_eval_pts). Per ognuno di questi punti abbiamo valutato la media e la deviazione standard dal GP e ricavato questi grafici
 
gp_prior
 
gp_posterior

Avevamo pensato di fare un confronto tra i grafici della posterior dopo 10,50,100 iterazioni oltre al confronto tra prior e posterior. Secondo te può essere uno spunto interessante?

**Comment by brunoguindani** [comment]
Mi sembra una buona idea! Però attenti quando scegliete una sola componente di uno spazio 30-dimensionale, è facile pensare che abbiate fatto cherry-picking. Per rimediare, potete mettere in presentazione una griglia con grafici per tutte le componenti (se è sufficientemente visibile), oppure solo alcune (possibilmente almeno 2 appunto) dicendo che le altre non sono rilevanti (e comunque nel report le mettete tutte).

In generale, non fate i salti mortali per rappresentare a tutti i costi una funzione di dimensione alta, è intrinsecamente difficile e non avete bisogno di giustificarvi. Una come esempio va bene, ma sicuramente quelle di dimensione più bassa sono più efficaci per comunicare i risultati

**Comment by simonegiacomello** [comment]
Abbiamo spostato il tutto su uno script separato che fa il plotting della prior e delle posterior dopo 10, 30, 50 iterazioni dell'algoritmo. L'unica funzione che abbiamo usato che presenta solo una componente X è la target Schaffer, quindi abbiamo deciso di riferirci solo a quella inserendo accoppiati i grafici delle due componenti obiettivo. 
I grafici sono questi. La diminuzione della varianza è evidente, il problema è che la curva è necessariamente una spezzata e presenta questi picchi che è impossibile prevedere data la casualità sia dei punti generati che delle funzioni che estraiamo dal GP. 

gp_schaffer_prior
gp_schaffer_iter=10_posterior
gp_schaffer_iter=30_posterior
gp_schaffer_iter=50_posterior

File: mobopt/_target_space.py [hunk 1]
```diff
@@ -354,4 +355,66 @@ def lhs_points(self, n_pts):
...
+            X = np.append(X, self._X, axis=0)
+
+        else:
+            raise ValueError("title must end with 'prior' or 'posterior'")
+
+        for i in range(self.NObj):
+            y_mean[i], y_std[i] = gpr_model[i].predict(X, return_std=True)
+            y_samples[i] = gpr_model[i].sample_y(X, n_samples=n_samples)
+
+        fig, ax = plt.subplots(1, 2, figsize=(20, 10))
```
Comment (brunoguindani) [inline mobopt/_target_space.py]: I parametri di questa griglia dovrebbero dipendere da `n_samples`, anche la `figsize`

File: mobopt/_target_space.py [hunk 2]
```diff
@@ -354,4 +355,66 @@ def lhs_points(self, n_pts):
...
+
+        for dim in range(self.NObj):
+
+            for idx in range(n_samples):
+                ax[dim].plot(
+                    np.sort(X[:,0]),
+                    -y_samples[dim, np.argsort(X[:, 0]), idx],
```
Comment (brunoguindani) [inline mobopt/_target_space.py]: Potreste salvarvi gli output di `sort` e `argsort` in variabili così non dovete chiamarle ogni volta e rendete il codice più leggibile

File: mobopt/_target_space.py [hunk 3]
```diff
@@ -354,4 +355,66 @@ def lhs_points(self, n_pts):
         for i in range(self.NParam):
             X[:, i] = X[:, i] * (self.pbounds[i][1] - self.pbounds[i][0]) + self.pbounds[i][0]
 
-        return X
\ No newline at end of file
+        return X
+
+    def plot_gp(self, gpr_model, n_eval_pts, title, n_samples=2):
```
Comment (brunoguindani) [inline mobopt/_target_space.py]: Avete modo di controllare il random seed che utilizzate in questa funzione? Mi sembra impoortante per ragioni di riproducibilità

**Review by brunoguindani — COMMENTED** [review]
Qualche consiglio sui grafici: anzitutto aggiungete `bbox_inches='tight'` a `savefig()`, così eiliminate un sacco di spazio bianco che rimpicciolisce la figura effettiva. Poi aumentate la dimensione del font, che al momento non è molto leggibile.

Qui sotto c'è anche qualche commento sul codice, ma direi che per quanto riguarda il materiale della presentazione siete a buon punto

File: mobopt/_target_space.py (hunk 1 above)
Comment (simonegiacomello) [reply mobopt/_target_space.py]: Non capisco il perchè dici che debbano dipendere da n_samples, che è il numero di funzioni che noi estraiamo dal GP e che plottiamo nello stesso grafico insieme alla media

File: mobopt/_target_space.py (hunk 3 above)
Comment (simonegiacomello) [reply mobopt/_target_space.py]: Se scegliamo di generare i punti con lhs no, mentre se generiamo con random points viene usata sempre la seed con cui abbiamo inizializzato la classe. Siccome i risultati non sembrano molto diversi, a questo punto correggo con random points.

Comment (simonegiacomello) [reply mobopt/_target_space.py]: Questo riguarda comunque solo la generazione dei punti, per i samples dal gp non si pone il problema perchè è definito da una sua seed autonoma

**Comment by simonegiacomello** [comment]
Abbiamo apportato le modifiche sulle variabili ripetute, abbiamo fissato il seed su tutto il codice uguale a RandomState che è un parametro di input, e modificato il rendering dei grafici. 
A meno che ci sia qualcosa che non va, la nostra intenzione sarebbe quella di chiudere qua con il codice. Facci sapere cosa ne pensi!

File: mobopt/_target_space.py (hunk 1 above)
Comment (brunoguindani) [reply mobopt/_target_space.py]: Mi sono sbagliato, intendevo `NObj` e non `n_samples`. L'osservazione comunque rimane: se plottate una funzione con output 3D dovreste avere 3 subplots e aggiustare la dimensione adeguatamente

File: mobopt/_target_space.py (hunk 3 above)
Comment (brunoguindani) [reply mobopt/_target_space.py]: Parlavo proprio dei samples dal GP, e della possibilità di settare il seed specificatamente in questa funzione, non code-wide. Quello che fate ora mi sembra potenzialmente pericoloso: state settando lo stesso random seed in due punti diversi del codice, perché la libreria lo faceva già [qui](https://github.com/simonegiacomello/MOBOpt/blob/review/mobopt/_target_space.py#L25).

Se riuscite a settare facilmente il seed dei GP in questa funzione (usando un argomento dello script che poi passate alla funzione) fatelo, se è troppo lavoro lasciate perdere. In ogni caso, togliete il `np.random.seed()` che avete aggiunto

**Review by brunoguindani — COMMENTED** [review]
Se sistemate le minor che noto qua sotto, per me il codice va anche bene così!

**Comment by simonegiacomello** [comment]
Abbiamo settato il seed direttamente nel GP e tolto la riga pericolosa. In più ora verranno stampati NObj grafici.

**Comment by brunoguindani** [comment]
Credo che non abbiate capito. Avevo detto:

> Se riuscite a settare facilmente il seed dei GP ***in questa funzione*** (usando un argomento dello script che poi passate alla funzione)

riferito alla funzione `plot_gp()`, non di settare il seed dei GP all'inizio del codice (che è quello che fate ora). La ragione è che voglio poter controllare i GP samples che finiscono nei grafici, non forzare allo stesso seed tutti i GP fin dall'inizio dell'ottimizzazione. Potete quindi aggiungere un argomento allo script `posterior_plots.py` e passarlo a `plot_gp()`. Stando alla [doc](https://scikit-learn.org/stable/modules/generated/sklearn.gaussian_process.GaussianProcessRegressor.html#sklearn.gaussian_process.GaussianProcessRegressor.sample_y), potete poi passare l'argomento `random_state` alla chiamata di `sample_y()` alla riga 387 di `_target_space.py`. Magari passate il seed + `i` invece che il seed e basta, così le varie chiamate di `sample_y()` nel loop hanno seed diversi. Togliete anche il setting in fase di inizializzazione dei GPR che avete appena aggiunto (quello a riga 177 di `_bayes.py`), e siamo a posto.

**Comment by simonegiacomello** [comment]
Ok abbiamo corretto. Non si poteva passare direttamente come parametro i + randomseed perchè randomseed era in ingresso al costruttore e non era possibile richiamarlo, in più self.RS non è un intero quindi non ci si poteva sommare i. 
Dunque abbiamo definito self.seed che è semplicemente randomseed e passato a y_sample random_state = i + self.seed

**Comment by brunoguindani** [comment]
Non ho capito perché ciò che suggerivo io non si possa fare. Vi ho raccomandato di aggiungere un argomento (da riga di comando, tipo `--target`) allo script `posterior_plots.py`, argomento che è quindi un intero. Passate poi questo intero alla funzione `plot_gp()` (aggiungendo un nuovo argomento), la quale a sua volta la passa a `sample_y()` dopo aver aggiunto `i`. Non vedo il problema, anche perché non ho mai parlato di `self.RS`. Ancora una volta, l'intenzione è che il seed sia un parametro di `plot_gp()` e che possa quindi essere deciso da chi la chiama. Adesso è vincolato ad essere uguale a quello dentro l'ottimizzatore!

A pensarci bene, visto che anche l'ottimizzatore in `posterior_plots.py` ha bisogno di un random seed (che ora settate sempre a 10), potete fare un unico argomento da riga di comando `--seed` (con il valore di default che volete). A riga 47 (cioè all'ottimizzatore) passate direttamente quel `seed`, mentre alle chiamate di `plot_gp()` di righe 52-57 passate `seed` + qualcosa. In questo modo, non usate lo stesso identico seed nelle due fasi di `posterior_plots.py`, ma lasciate libero l'utente di dare a `plot_gp()` il seed che preferisce se volesse chiamarla in altri script.

Un'ultima cosa: dopo aver sistemato, scrivetemi per cortesia gli esatti comandi che avete usato per ottenere i vostri ultimi plot, così controllo che funzioni tutto

**Comment by simonegiacomello** [comment]
Scusami non avevo capito, comunque per essere consistenti abbiamo aggiunto l'argomento seed a tutti gli script con default 10 che era il valore iniziale.
Per eseguire lo script posterior_plots.py basta questa riga
--target SCHAFFER -np 20

Salverà i grafici della prior e delle posterior dopo 10, 30 e 50 iterazioni come png nella cartella scripts.

**Review by brunoguindani — APPROVED** [review]
Bene, tutto funziona! Faccio il merge
